packages/lesswrong/lib/sql/InsertQuery.ts

```diff
@@ -0,0 +1,136 @@
+import Query, { Atom } from "./Query";
+import Table from "./Table";
+import { Type } from "./Type";
+import { randomId } from '../random';
+
+export type ConflictStrategy = "error" | "ignore" | "upsert";
+
+type InsertSqlBaseOptions = Partial<{
+  returnInserted: boolean,
+}>
+
+export type InsertSqlConflictOptions<T extends DbObject> = Partial<{
+  conflictStrategy: Omit<ConflictStrategy, "upsert">,
+  upsertSelector: never,
+}> | {
+  conflictStrategy: "upsert",
+  upsertSelector?: MongoSelector<T>,
+}
+
+export type InsertSqlOptions<T extends DbObject> = InsertSqlBaseOptions & InsertSqlConflictOptions<T>;
+
+/**
+ * Builds a Postgres query to insert some specific data into the given table.
+ *
+ * `data` may be either a single collection object to insert, or an array of
+ * objects.
+ *
+ * By default, a primary key conflict will result in an error being thrown when
+ * the query is executed. To change this behaviour, `sqlOptions` may be used to
+ * either ignore the conflict, or to enable upserting.
+ *
+ * When upserting, you can also supply an optional `upsertSelector` to test for
+ * conflicts on fields other than the primary key, but note that, like Mongo, a
+ * unique index must exist on exactly those fields to work. In general, this
+ * should be considered fragile; 1) because of the requirement for a unique index
+ * (meaning that indexes define the behaviour of the query, not just the speed), and
+ * 2) because Mongo and Postgres have different opinions on how conflicts should be
+ * treated on null fields which means we have to do some magic with `COALESCE` to make
+ * things backwards compatible (see CreateIndexQuery). If you use upserting, test it
+ * well!
+ */
+class InsertQuery<T extends DbObject> extends Query<T> {
+  constructor(
+    table: Table,
+    data: T | T[],
+    _options: MongoInsertOptions<T> = {}, // TODO: What can options be?
+    sqlOptions?: InsertSqlOptions<T>,
+  ) {
+    super(table, [`INSERT INTO "${table.getName()}"`]);
+    data = Array.isArray(data) ? data : [data];
```

Comment:
nit: to get the `T[]` type info
```suggestion
    const dataAsArray = Array.isArray(data) ? data : [data];
```